rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,347 @@
+# RFC 9811 - 2021-11-12 - VRL enum VM (VenuM)
+
+This RFC proposes implementing a enum VM. VRL will be compiled to a list of
+instructions and executed by this VM, with the aim to significantly improve the
+performance of executing VRL programs.
+
+## Context
+
+This RFC is a follow on from
+[rfcs/2021-10-14-9811-vrl-performance.md](https://github.com/vectordotdev/vector/pull/9812)
+
+## Scope
+
+### In scope
+
+This RFC is purely about developing a enum VM for VRL. It will discuss the
+risks involved in running VRL with a VM and the measures we will take to
+mitigate those risks.
+
+### Out of scope
+
+Any other performance issues relating to VRL will be discussed in
+[rfcs/2021-10-14-9811-vrl-performance.md](https://github.com/vectordotdev/vector/pull/9812)
+
+## Pain
+
+Vrl compiles to an AST that is then walked during resolution. Each node in that
+tree is boxed and stored in disparate regions of memory. As a result walking
+the tree means that the CPU caches must be constantly swapped.
+
+## Proposal
+
+### Implementation
+
+Instead we can create an enum VM to store the execution of the Vrl program.
+
+The enum is essentially a big enum of instructions:
+
+```rust
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum OpCode {
+    Return,
+    Constant,
+    Negate,
+    Add,
+    Subtract,
+    Multiply,
+    Divide,
+    Print,
+    Not,
+    Greater,
+    GreaterEqual,
+    Less,
+    LessEqual,
+    NotEqual,
+    Equal,
+    Pop,
+    JumpIfFalse,
+    Jump,
+    SetPath,
+    GetPath,
+    Call,
+    ...
+}
+
+enum Instruction {
+    Opcode(Opcode),
+    Literal(LiteralIndex),
+}
```

Comment:
I know we've discussed reworking this to put the literals into the opcode enum, but I just realized another disadvantage of this format: each opcode that takes a literal takes a minimum of four words this way, making it dramatically less efficient for storage for no real gain. We would be better to use the other approach (literals into the opcode enum) with two words per opcode, or an untyped literal byte code, but I suspect the former would be simpler to implement.